### PR TITLE
Add S3 Select support for Parquet using S3-SELECT=AUTO

### PR DESCRIPTION
**Pull request by frankgh**
Fix CSV serialization when the input has special characters (NEWLINE,
DELIMITER, QUOTE).

Co-authored-by: Raymond Yin <ryin@pivotal.io>
Co-authored-by: Francisco Guerrero <aguerrero@pivotal.io>

### CODE REVIEW

File: server/pxf-api/src/main/java/org/greenplum/pxf/api/GreenplumCSV.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package org.greenplum.pxf.api;
+
+/**
+ * Greenplum CSV Default
+ */
+public class GreenplumCSV {
```
Comment (denalex) [inline server/pxf-api/src/main/java/org/greenplum/pxf/api/GreenplumCSV.java]: if just for constants, this can be an interface

File: server/pxf-api/src/main/java/org/greenplum/pxf/api/utilities/Utilities.java [hunk 2]
```diff
@@ -56,7 +60,7 @@
     public static String toCsvText(String s,
                                    boolean prependQuoteChar,
                                    boolean appendQuoteChar) {
-        return toCsvText(s, CSV_QUOTE, prependQuoteChar, appendQuoteChar, false);
+        return toCsvText(s, QUOTE, ESCAPE, NEWLINE, DELIMITER, prependQuoteChar, appendQuoteChar, false);
```
Comment (denalex) [inline server/pxf-api/src/main/java/org/greenplum/pxf/api/utilities/Utilities.java]: might as well pass in a POJO that encapsulates the details for a given CSV definition, with `GreenpluDefaultCSV` being and singleton POJO for defaults. But also need to limit object creation, so for a given request a single POJO can be created (and stored in the RequestContext) so that each row transformation utilizes the same object.

File: server/pxf-api/src/main/java/org/greenplum/pxf/api/utilities/Utilities.java [hunk 3]
```diff
@@ -73,24 +77,47 @@ public static String toCsvText(String s,
      */
     public static String toCsvText(String s,
```
Comment (denalex) [inline server/pxf-api/src/main/java/org/greenplum/pxf/api/utilities/Utilities.java]: I have hard time believing this problem has not been solved already a thousand times. Is there any standard library that does CSV conversion ?

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/BridgeOutputBuilder.java [hunk 4]
```diff
@@ -444,8 +446,15 @@ else if (field.type == DataType.TIMESTAMP.getOID())
...
+                                QUOTE,
+                                QUOTE,
+                                NEWLINE,
+                                DELIMITER,
+                                true,
+                                true,
+                                true);
                 })
-                .collect(Collectors.joining(DELIMITER, "", NEWLINE));
+                .collect(Collectors.joining(DELIMITER + "", "", NEWLINE));
```
Comment (denalex) [inline server/pxf-service/src/main/java/org/greenplum/pxf/service/BridgeOutputBuilder.java]: what's the point of adding an empty string to the delimiter ?

File: server/pxf-api/src/main/java/org/greenplum/pxf/api/utilities/Utilities.java [hunk 5]
```diff
@@ -73,24 +77,47 @@ public static String toCsvText(String s,
...
                                    char quoteChar,
+                                   char escapeChar,
```
Comment (lavjain) [inline server/pxf-api/src/main/java/org/greenplum/pxf/api/utilities/Utilities.java]: toCSVText function should belong to GreenplumCSV class

**Comment by lavjain** [comment]
The following function (line 261) must be changed to use a static list.

```
boolean isStringType(DataType type)
```

File: server/pxf-api/src/main/java/org/greenplum/pxf/api/utilities/Utilities.java (hunk 2 above)
Comment (lavjain) [reply server/pxf-api/src/main/java/org/greenplum/pxf/api/utilities/Utilities.java]: QUOTE, ESCAPE, NEWLINE, DELIMITER, etc should be properties of GreenplumCSV object

File: server/pxf-service/src/main/java/org/greenplum/pxf/service/BridgeOutputBuilder.java (hunk 4 above)
Comment (frankgh) [inline server/pxf-service/src/main/java/org/greenplum/pxf/service/BridgeOutputBuilder.java]: Let's use String.valueOf(DELIMITER) or define a DELIMITER_STRING
```suggestion
                .collect(Collectors.joining(String.valueOf(DELIMITER), "", NEWLINE));
```

File: server/pxf-s3/src/main/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandler.java [hunk 6]
```diff
@@ -25,14 +25,9 @@
             new SupportMatrixEntry("TEXT", OutputFormat.TEXT, S3Mode.ON),
             new SupportMatrixEntry("CSV", OutputFormat.TEXT, S3Mode.ON),
             new SupportMatrixEntry("JSON", OutputFormat.TEXT, S3Mode.ON),
-            new SupportMatrixEntry("PARQUET", OutputFormat.TEXT, S3Mode.AUTO),
-            new SupportMatrixEntry("TEXT", OutputFormat.TEXT, S3Mode.AUTO),
-            new SupportMatrixEntry("CSV", OutputFormat.TEXT, S3Mode.AUTO),
-            new SupportMatrixEntry("JSON", OutputFormat.TEXT, S3Mode.AUTO),
+            new SupportMatrixEntry("PARQUET", OutputFormat.TEXT, S3Mode.OFF),
```
[the remark this answers is not in the document]
Comment (shivzone) [reply server/pxf-s3/src/main/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandler.java]: Other question, does compression affect s3 select support ?

[the remark this answers is not in the document]
Comment (raymondyin) [reply server/pxf-s3/src/main/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandler.java]: @shivzone not necessarily. We did not support Parquet using TEXT or CSV format, only FORMAT='custom' was supported. 

For your second question, compression should affect s3-select support. We'll write a follow up story for compression when S3-SELECT=AUTO is used

File: server/pxf-api/src/main/java/org/greenplum/pxf/api/model/GreenplumCSV.java [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+            if (delimiter != null && curr == delimiter) specialChars++;
+            if (newline != null && newline.length() > 0) {
+
+                j = 0;
+
+                // let's say we have input asd\r\nacd
+                // and newline \r\n then we need to
+                // increase the specialChars count by 1
+
+                while (i < length && j < newline.length()
```
Comment (oliverralbertini) [inline server/pxf-api/src/main/java/org/greenplum/pxf/api/model/GreenplumCSV.java]: this would seem to warrant tests for edge cases like newlines at the beginning/end/middle of a string?

File: server/pxf-api/src/test/java/org/greenplum/pxf/api/model/GreenplumCSVTest.java [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+        assertEquals(quoteChar + expected, gpCSV.toCsvText(input, true, false, false));
+        assertEquals(expected, gpCSV.toCsvText(input, false, false, false));
+        assertEquals(expected + quoteChar, gpCSV.toCsvText(input, false, true, false));
+        assertEquals(quoteChar + expected + quoteChar,
+                gpCSV.toCsvText(input, true, true, false));
+    }
+
+    @Test
+    public void testWithValueOfNullValidDefaultInput() {
```
Comment (oliverralbertini) [inline server/pxf-api/src/test/java/org/greenplum/pxf/api/model/GreenplumCSVTest.java]: It's confusing to name all the tests with `*Input()`, can we use a different word to reflect that these are CSV options? Especially since the text itself is often named `input`.

File: server/pxf-s3/src/main/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandler.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (oliverralbertini) [reply server/pxf-s3/src/main/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandler.java]: I agree we need the `OFF` mode in the matrix, glad we got rid of the `AUTO` one.

File: server/pxf-s3/src/main/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandler.java [hunk 9]
```diff
@@ -87,19 +81,19 @@ private boolean useS3Select(RequestContext context) {
 
         switch (selectMode) {
             case ON:
-                return formatSupported(outputFormat, format, S3Mode.ON, compressionType, true);
```
Comment (oliverralbertini) [inline server/pxf-s3/src/main/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandler.java]: I'm unclear on why we got rid of `compressionType` here.

File: server/pxf-s3/src/main/java/org/greenplum/pxf/plugins/s3/S3SelectFilterParser.java [hunk 10]
```diff
@@ -142,7 +148,7 @@ private String mapValue(Object val, DataType dataType) {
             case TEXT:
             case VARCHAR:
             case BPCHAR:
-                return Utilities.toCsvText(val.toString(), '\'', true, true, false);
+                return SINGLE_QUOTE_ESCAPE.toCsvText(val.toString(), true, true);
```
Comment (oliverralbertini) [inline server/pxf-s3/src/main/java/org/greenplum/pxf/plugins/s3/S3SelectFilterParser.java]: I know this is just following the previous pattern, but it seems like an abuse of the `GreenplumCSV` class to use it when parsing the filter to prepare a string that represents a SQL query *and* when reading data from S3. I think it should still be used, but maybe we can come up with a better name since this is not strictly CSV. Maybe `GreenplumCsvText`?

File: server/pxf-s3/src/test/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandlerTest.java [hunk 11]
```diff
@@ -28,7 +28,7 @@
 
     private static String[] EXPECTED_RESOLVER_TEXT_ON = {STRING_PASS_RESOLVER, STRING_PASS_RESOLVER, STRING_PASS_RESOLVER, STRING_PASS_RESOLVER, NOT_SUPPORTED};
     private static String[] EXPECTED_RESOLVER_TEXT_AUTO_BENEFIT = {STRING_PASS_RESOLVER, STRING_PASS_RESOLVER, STRING_PASS_RESOLVER, STRING_PASS_RESOLVER, DEFAULT_RESOLVER};
-    private static String[] EXPECTED_RESOLVER_TEXT_AUTO_NO_BENEFIT = {STRING_PASS_RESOLVER, DEFAULT_RESOLVER, DEFAULT_RESOLVER, STRING_PASS_RESOLVER, DEFAULT_RESOLVER};
+    private static String[] EXPECTED_RESOLVER_TEXT_AUTO_NO_BENEFIT = {DEFAULT_RESOLVER, DEFAULT_RESOLVER, DEFAULT_RESOLVER, STRING_PASS_RESOLVER, DEFAULT_RESOLVER};
```
Comment (oliverralbertini) [inline server/pxf-s3/src/test/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandlerTest.java]: @raymondyin and I tried to break down why we want the defaults for Fragmenter, Accessor, Resolver, but couldn't come up with an explanation. Was this a mistake, or does this PR introduce changes that make this necessary?

File: server/pxf-s3/src/test/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandlerTest.java [hunk 12]
```diff
@@ -46,13 +46,18 @@
...
     private static String[] EXPECTED_ACCESSOR_TEXT_OFF = {DEFAULT_ACCESSOR, DEFAULT_ACCESSOR, DEFAULT_ACCESSOR, DEFAULT_ACCESSOR, DEFAULT_ACCESSOR};
 
     private static String[] EXPECTED_ACCESSOR_GPDB_WRITABLE_ON = {NOT_SUPPORTED, NOT_SUPPORTED, NOT_SUPPORTED, NOT_SUPPORTED, NOT_SUPPORTED};
     private static String[] EXPECTED_ACCESSOR_GPDB_WRITABLE_AUTO = {DEFAULT_ACCESSOR, NOT_SUPPORTED, NOT_SUPPORTED, DEFAULT_ACCESSOR, DEFAULT_ACCESSOR};
     private static String[] EXPECTED_ACCESSOR_GPDB_WRITABLE_OFF = {DEFAULT_ACCESSOR, DEFAULT_ACCESSOR, DEFAULT_ACCESSOR, DEFAULT_ACCESSOR, DEFAULT_ACCESSOR};
 
+    private static String[] EXPECTED_ACCESSOR_TEXT_AUTO_NO_BENEFIT_HAS_FORMAT_OPTIONS = {DEFAULT_ACCESSOR, S3_ACCESSOR, S3_ACCESSOR, S3_ACCESSOR, DEFAULT_ACCESSOR};
+    private static String[] EXPECTED_RESOLVER_TEXT_AUTO_NO_BENEFIT_HAS_FORMAT_OPTIONS = {DEFAULT_RESOLVER, STRING_PASS_RESOLVER, STRING_PASS_RESOLVER, STRING_PASS_RESOLVER, DEFAULT_RESOLVER};
+    private static String[] EXPECTED_FRAGMENTER_TEXT_AUTO_NO_BENEFIT_HAS_FORMAT_OPTIONS = {DEFAULT_FRAGMENTER, FILE_FRAGMENTER, FILE_FRAGMENTER, FILE_FRAGMENTER, DEFAULT_FRAGMENTER};
```
Comment (oliverralbertini) [inline server/pxf-s3/src/test/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandlerTest.java]: This test is super hard to understand!

Comment (oliverralbertini) [reply server/pxf-s3/src/test/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandlerTest.java]: Why are we using `S3_ACCESSOR` when there is no benefit to do so when there are format options present?

File: server/pxf-s3/src/main/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandler.java (hunk 9 above)
Comment (frankgh) [reply server/pxf-s3/src/main/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandler.java]: There's a story in the backlog to add it back. It should be simple as in we need a map where the key is the format and the value is a set of supported formats. Then we would need a check similar to this

https://github.com/greenplum-db/pxf/blob/6fa8208c9215c2a733a0498c91a1bf03077c86e5/server/pxf-s3/src/main/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandler.java#L73-L80

Comment (frankgh) [reply server/pxf-s3/src/main/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandler.java]: maybe it should be part of this story actually

File: server/pxf-s3/src/main/java/org/greenplum/pxf/plugins/s3/S3SelectFilterParser.java (hunk 10 above)
Comment (frankgh) [reply server/pxf-s3/src/main/java/org/greenplum/pxf/plugins/s3/S3SelectFilterParser.java]: it's ugly, I agree. I feel StringUtils.replace should be fine here

File: server/pxf-s3/src/test/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandlerTest.java (hunk 12 above)
Comment (frankgh) [reply server/pxf-s3/src/test/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandlerTest.java]: it's because the format options only affect the csv/text format, and not the parquet format

Comment (frankgh) [reply server/pxf-s3/src/test/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandlerTest.java]: the tests are hard to understand. I agree

File: server/pxf-api/src/test/java/org/greenplum/pxf/api/model/GreenplumCSVTest.java (hunk 8 above)
Comment (raymondyin) [reply server/pxf-api/src/test/java/org/greenplum/pxf/api/model/GreenplumCSVTest.java]: Test method names changed.

File: server/pxf-s3/src/main/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandler.java (hunk 9 above)
Comment (raymondyin) [reply server/pxf-s3/src/main/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandler.java]: Added compression type logic

File: server/pxf-s3/src/main/java/org/greenplum/pxf/plugins/s3/S3SelectFilterParser.java (hunk 10 above)
Comment (raymondyin) [reply server/pxf-s3/src/main/java/org/greenplum/pxf/plugins/s3/S3SelectFilterParser.java]: Updated parsing logic for this case.

File: server/pxf-s3/src/test/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandlerTest.java (hunk 11 above)
Comment (raymondyin) [reply server/pxf-s3/src/test/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandlerTest.java]: Francisco explained that when there's no benefit, we use default resolver for "parquet", "text", "csv", "avro", and string pass resolver for "json".

File: server/pxf-s3/src/test/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandlerTest.java [hunk 13]
```diff
@@ -380,6 +380,42 @@ public void testSelectAutoUnsupportedFormat() {
...
+        context.addOption(S3SelectAccessor.COMPRESSION_TYPE, "foo");
+        // EXPECTED_RESOLVER_GPDB_WRITABLE_ON is used as its values are desirable as expected value
+        verifyResolvers(context, EXPECTED_RESOLVER_GPDB_WRITABLE_ON);
+    }
+
+    @Test
+    public void testWithSelectOffAndWithUnSupportedCompressionType() {
+        context.addOption("S3-SELECT", "off");
+        context.addOption(S3SelectAccessor.COMPRESSION_TYPE, "foo");
+        // EXPECTED_RESOLVER_GPDB_WRITABLE_ON is used as its values are desirable as expected value
```
Comment (oliverralbertini) [inline server/pxf-s3/src/test/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandlerTest.java]: you mean `EXPECTED_RESOLVER_GPDB_WRITABLE_OFF`?

Comment (raymondyin) [reply server/pxf-s3/src/test/java/org/greenplum/pxf/plugins/s3/S3ProtocolHandlerTest.java]: Good catch! Correct.

Thank you Oliver.
